pkg/v3/ocr3.go

```diff
@@ -0,0 +1,60 @@
+package v3
+
+import (
+	"context"
+
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+)
+
+type InstructionStore interface{}
+
+// SamplingStore at this point can be an empty interface
+type SamplingStore interface{}
+
+// ResultStore at this point can be an empty interface
+type ResultStore interface{}
+
+type ocr3Plugin struct {
+	PrebuildHooks     []func(AutomationOutcome) error
+	BuildHooks        []func(*AutomationObservation, InstructionStore, SamplingStore, ResultStore) error
+	InstructionSource InstructionStore
+	MetadataSource    SamplingStore
+	ResultSource      ResultStore
+}
+
+func (plugin *ocr3Plugin) Observation(ctx context.Context, outcome ocr3types.OutcomeContext, query types.Query) (types.Observation, error) {
+	// Decode the outcome to AutomationOutcome
+	automationOutcome, err := DecodeAutomationOutcome(outcome.PreviousOutcome)
+	if err != nil {
+		return nil, err
+	}
+
+	// Execute pre-build hooks
+	for _, hook := range plugin.PrebuildHooks {
+		err := hook(automationOutcome)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
I was just now thinking about the error handling here. It's similar to something I was working with today where the question is do you exit early on an error or run all hooks and returned a joined error? In my case, I went with the latter and returned a joined error after the for loop. That would indicate that the pre-build process failed but allows all parts to run.